File: src/test/java/com/raychenon/leetcode/QueensThatCanAttackTheKingTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static junit.framework.TestCase.assertTrue;

/**
 * User: raychenon
 * Date: 2019-11-30
 * https://leetcode.com/problems/queens-that-can-attack-the-king/
 */
public class QueensThatCanAttackTheKingTest {

    private static QueensThatCanAttackTheKing queensThatCanAttackTheKing;

    @BeforeClass
    public static void setUp() {
        queensThatCanAttackTheKing = new QueensThatCanAttackTheKing();
    }


    // |K|A| | |Q| | | |
    // |A| | | | | | | |
    // | | | | |Q| | | |
    // | | | |A| | | | |
    // |Q| | | | | | | |
    // | | | | | | | | |
    // | | | | | | | | |
    // | | | | | | | | |
    @Test
    public void queens1Test() {
        int[][] queens = new int[][]{
                {0, 1}, {1, 0}, {4, 0}, {0, 4}, {3, 3}, {2, 4}
        };

        int[] king = new int[]{0, 0};

        int[][] expected = new int[][]{
                {0, 1}, {1, 0}, {3, 3}
        };

        compareMethods(expected, queens, king);
    }

    // |Q| | | | | | | |
    // | |Q| | | | | | |
    // | | |A| | | | | |
    // | | | |K|A|Q| | |
    // | | | | |A|Q| | |
    // | | | | | | | | |
    // | | | | | | | | |
    // | | | | | | | | |
    @Test
    public void queens2Test() {
        int[][] queens = new int[][]{
                {0, 0}, {1, 1}, {2, 2}, {3, 4}, {3, 5}, {4, 4}, {4, 5}
        };

        int[] king = new int[]{3, 3};

        int[][] expected = new int[][]{
                {2, 2}, {3, 4}, {4, 4}
        };

        compareMethods(expected, queens, king);
    }


    // |Q|Q| |Q|Q|Q|Q|Q|
    // |Q|Q|Q| |A| |A|Q|
    // | |Q|Q|A| | |Q|Q|
    // | | | | |K| | |A|
    // |Q| |Q|A| |A|Q| |
    // |Q|Q|Q| |A| |Q| |
    // | |Q| |Q|Q| | | |
    // | | | | | |Q| |Q|
    @Test
    public void queens3Test() {
        int[][] queens = new int[][]{
                {5, 6}, {7, 7}, {2, 1}, {0, 7}, {1, 6}, {5, 1}, {3, 7},
                {0, 3}, {4, 0}, {1, 2}, {6, 3}, {5, 0}, {0, 4}, {2, 2},
                {1, 1}, {6, 4}, {5, 4}, {0, 0}, {2, 6}, {4, 5}, {5, 2},
                {1, 4}, {7, 5}, {2, 3}, {0, 5}, {4, 2}, {1, 0}, {2, 7},
                {0, 1}, {4, 6}, {6, 1}, {0, 6}, {4, 3}, {1, 7}
        };

        int[] king = new int[]{3, 4};

        int[][] expected = new int[][]{
                {2, 3}, {1, 4}, {1, 6}, {3, 7}, {4, 3}, {5, 4}, {4, 5}
        };

        compareMethods(expected, queens, king);
    }


    public void compareMethods(int[][] expectedQueens, int[][] queens, int[] king) {

        Set<QueenCoordinate> expectedQueensSubset = new HashSet<>();
        for (int[] queen : expectedQueens) {
            QueenCoordinate coord = new QueenCoordinate(queen[0], queen[1]);
            expectedQueensSubset.add(coord);
        }

        Set<QueenCoordinate> actualQueensSubset = new HashSet<>();
        for (List<Integer> queenCoordinates : queensThatCanAttackTheKing.queensAttacktheKing(queens, king)) {
            QueenCoordinate temp = new QueenCoordinate(queenCoordinates.get(0), queenCoordinates.get(1));
            actualQueensSubset.add(temp);
        }

        assertTrue(actualQueensSubset.containsAll(expectedQueensSubset));
        assertTrue(expectedQueensSubset.equals(actualQueensSubset));
    }


    class QueenCoordinate {
        private int row;
        private int col;

        public QueenCoordinate(int y, int x) {
            this.row = y;
            this.col = x;
        }

        @Override
        public boolean equals(Object o) {

            if (o == this) return true;
            if (!(o instanceof QueenCoordinate)) {
                return false;
            }

            QueenCoordinate user = (QueenCoordinate) o;

            return user.row == row &&
                    user.col == col;
        }

        @Override
        public int hashCode() {
            int result = 17;
            result = 31 * result + row;
            result = 31 * result + col;
            return result;
        }
    }
}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/QueensThatCanAttackTheKingTest.kt
package com.raychenon.kotlin.leetcode

import junit.framework.TestCase
import org.junit.Test
import java.util.*

/**
 * User: raychenon
 * Date: 2019-12-01
 * https://leetcode.com/problems/queens-that-can-attack-the-king/
 */
class QueensThatCanAttackTheKingTest {


    // |K|A| | |Q| | | |
    // |A| | | | | | | |
    // | | | | |Q| | | |
    // | | | |A| | | | |
    // |Q| | | | | | | |
    // | | | | | | | | |
    // | | | | | | | | |
    // | | | | | | | | |
    @Test
    fun queens1Test() {
        val queens = arrayOf(
            intArrayOf(0, 1),
            intArrayOf(1, 0),
            intArrayOf(4, 0),
            intArrayOf(0, 4),
            intArrayOf(3, 3),
            intArrayOf(2, 4)
        )
        val king = intArrayOf(0, 0)
        val expected = arrayOf(intArrayOf(0, 1), intArrayOf(1, 0), intArrayOf(3, 3))
        compareMethods(expected, queens, king)
    }

    // |Q| | | | | | | |
    // | |Q| | | | | | |
    // | | |A| | | | | |
    // | | | |K|A|Q| | |
    // | | | | |A|Q| | |
    // | | | | | | | | |
    // | | | | | | | | |
    // | | | | | | | | |
    @Test
    fun queens2Test() {
        val queens = arrayOf(
            intArrayOf(0, 0),
            intArrayOf(1, 1),
            intArrayOf(2, 2),
            intArrayOf(3, 4),
            intArrayOf(3, 5),
            intArrayOf(4, 4),
            intArrayOf(4, 5)
        )
        val king = intArrayOf(3, 3)
        val expected = arrayOf(intArrayOf(2, 2), intArrayOf(3, 4), intArrayOf(4, 4))
        compareMethods(expected, queens, king)
    }
    
    // |Q|Q| |Q|Q|Q|Q|Q|
    // |Q|Q|Q| |A| |A|Q|
    // | |Q|Q|A| | |Q|Q|
    // | | | | |K| | |A|
    // |Q| |Q|A| |A|Q| |
    // |Q|Q|Q| |A| |Q| |
    // | |Q| |Q|Q| | | |
    // | | | | | |Q| |Q|
    @Test
    fun queens3Test() {
        val queens = arrayOf(
            intArrayOf(5, 6),
            intArrayOf(7, 7),
            intArrayOf(2, 1),
            intArrayOf(0, 7),
            intArrayOf(1, 6),
            intArrayOf(5, 1),
            intArrayOf(3, 7),
            intArrayOf(0, 3),
            intArrayOf(4, 0),
            intArrayOf(1, 2),
            intArrayOf(6, 3),
            intArrayOf(5, 0),
            intArrayOf(0, 4),
            intArrayOf(2, 2),
            intArrayOf(1, 1),
            intArrayOf(6, 4),
            intArrayOf(5, 4),
            intArrayOf(0, 0),
            intArrayOf(2, 6),
            intArrayOf(4, 5),
            intArrayOf(5, 2),
            intArrayOf(1, 4),
            intArrayOf(7, 5),
            intArrayOf(2, 3),
            intArrayOf(0, 5),
            intArrayOf(4, 2),
            intArrayOf(1, 0),
            intArrayOf(2, 7),
            intArrayOf(0, 1),
            intArrayOf(4, 6),
            intArrayOf(6, 1),
            intArrayOf(0, 6),
            intArrayOf(4, 3),
            intArrayOf(1, 7)
        )
        val king = intArrayOf(3, 4)
        val expected = arrayOf(
            intArrayOf(2, 3),
            intArrayOf(1, 4),
            intArrayOf(1, 6),
            intArrayOf(3, 7),
            intArrayOf(4, 3),
            intArrayOf(5, 4),
            intArrayOf(4, 5)
        )
        compareMethods(expected, queens, king)
    }


    fun compareMethods(
        expectedQueens: Array<IntArray>,
        queens: Array<IntArray>,
        king: IntArray
    ) {
        val expectedQueensSubset: MutableSet<QueenCoordinate> = HashSet()
        for (queen in expectedQueens) {
            val coord = QueenCoordinate(queen[0], queen[1])
            expectedQueensSubset.add(coord)
        }
        val actualQueensSubset: MutableSet<QueenCoordinate> = HashSet()
        for (queenCoordinates in QueensThatCanAttackTheKing.queensAttacktheKing(
            queens,
            king
        )) {
            val temp = QueenCoordinate(queenCoordinates[0], queenCoordinates[1])
            actualQueensSubset.add(temp)
        }
        TestCase.assertTrue(actualQueensSubset.containsAll(expectedQueensSubset))
        TestCase.assertTrue(expectedQueensSubset == actualQueensSubset)
    }


    data class QueenCoordinate(val row: Int, val col: Int)
}